Adding AllenNLP Interpret

matt-gardner opened the pull request:
This is a greatly cleaned up version of #219. I refactored a ton of stuff to make the components self-contained, instead of needing to customize themselves based on the task.  Instead, each task does its own custom logic and passes it to the component, which has a single (customizable) code path for all tasks.

I have tested this locally with small models for (1) sentiment analysis, (2) textual entailment, (3) NER, (4) reading comprehension.  It appears to work with all of them.

I have not yet added the language modeling and masked language modeling demos, but this is already a very large PR.  I'll open a subsequent PR to add those two.

I'm not great with javascript or react, so it's quite possible I did some non-idiomatic things.  But this compiles without any new eslint warnings.

joelgrus commented on app.py:
```diff
@@ -175,6 +204,83 @@ def permadata(model_name: str) -> Response:  # pylint: disable=unused-variable
                 "responseData": permadata.response_data
         })
 
+    @app.route('/attack/<model_name>/<attacker_name>/<name_of_input_to_attack>/<name_of_grad_input>',
```
it feels weird to me to have all of these parameters as parts of the route, although I don't have a strong justification as why.

if I had implemented this, I most likely would have had the route just be `/attack/<model_name>`, with the remaining parameters as part of the body

matt-gardner replied:
I agree with this.  I assume I can look at how `/predict` works to refactor this?  I didn't really look closely at this part of the code.

joelgrus commented on app.py:
```diff
@@ -175,6 +204,83 @@ def permadata(model_name: str) -> Response:  # pylint: disable=unused-variable
+        if request.method == "OPTIONS":
+            return Response(response="", status=200)
+        lowered_model_name = model_name.lower()
+        attacker = app.attackers.get(lowered_model_name).get(attacker_name)
```
I think this will crash if the model name is not in app.attackers? should we be more defensive here?

joelgrus commented on app.py:
```diff
@@ -175,6 +204,83 @@ def permadata(model_name: str) -> Response:  # pylint: disable=unused-variable
+        attack = attacker.attack_from_json(data, name_of_input_to_attack, name_of_grad_input)
+        return jsonify(attack)
+
+    @app.route('/targeted_attack/<model_name>/<attacker_name>/<name_of_input_to_attack>/<name_of_grad_input>/<target>',
```
same comment as above

joelgrus commented on app.py:
```diff
@@ -175,6 +204,83 @@ def permadata(model_name: str) -> Response:  # pylint: disable=unused-variable
+        if request.method == "OPTIONS":
+            return Response(response="", status=200)
+        lowered_model_name = model_name.lower()
+        attacker = app.attackers.get(lowered_model_name).get(attacker_name)
```
same comment as above

joelgrus commented on app.py:
```diff
@@ -175,6 +204,83 @@ def permadata(model_name: str) -> Response:  # pylint: disable=unused-variable
+        if request.method == "OPTIONS":
+            return Response(response="", status=200)
+        lowered_model_name = model_name.lower()
+        print(app.interpreters)
```
remove print statement

joelgrus commented on demo/src/components/Hotflip.js:
```diff
@@ -0,0 +1,155 @@
+    } from 'react-accessible-accordion';
+import { RedToken, GreenToken, TransparentToken } from './Shared';
+import {  HOTFLIP_ATTACKER } from './InterpretConstants'
+//import { transformToTree } from './demos/Coref'
```
remove commented out code

matt-gardner commented:
Also, one thing I would like, but wasn't really sure how to implement, is some kind of "waiting" indicator when you push the "interpret" or "attack" buttons.  Any pointers on how to do that?

joelgrus commented on demo/src/components/Hotflip.js:
```diff
@@ -0,0 +1,155 @@
+            var original_string = ''
+            var flipped_string = ''
+            var new_prediction = ''
+            var context = " ";
```
might as well use `let` for these, since we're using modern JS everywhere else

joelgrus commented on demo/src/components/Hotflip.js:
```diff
@@ -0,0 +1,155 @@
+        const { hotflipData, hotflipFunction, requestDataObject, attacker, nameOfInputToAttack, nameOfGradInput } = this.props
+        if (attacker === HOTFLIP_ATTACKER){ // if attacker is not INPUT_REDUCTION or other methods
+            var original_string = ''
+            var flipped_string = ''
```
more idiomatic would be `originalString`, `flippedString`

joelgrus commented on demo/src/components/Hotflip.js:
```diff
@@ -0,0 +1,155 @@
+                new_prediction = hotflipData["new_prediction"]
+                context = hotflipData["context"]
+            }
+            const run_button = <button
```
more idiomatic `runButton` or even use styled components to make a `RunButton`

joelgrus commented on demo/src/components/InputReduction.js:
```diff
@@ -0,0 +1,124 @@
+    let original_string_colorized = []
+    let reduced_string_colorized = []
+    var idx = 0;
+    var idx2 = 0;
```
let

joelgrus commented on demo/src/components/InputReduction.js:
```diff
@@ -0,0 +1,124 @@
+        // if the words are not equal, keep traversing along the original list, making
+        // each token red as you go along. Also add a blank placeholder into the
+        // reduced list to make the words line up spacing wise in the UI.
+        else {
```
I believe it's more idiomatic to do 

```
} else {
```

cf https://github.com/airbnb/javascript#blocks--cuddled-elses

joelgrus commented on demo/src/components/InputReduction.js:
```diff
@@ -0,0 +1,124 @@
+                                  Reduce Input
+                               </button>
+
+            var display_text = '';
```
let

joelgrus replied:
and `displayText`

joelgrus commented on demo/src/components/InputReduction.js:
```diff
@@ -0,0 +1,124 @@
+                // (2) you can format the original input and the reduced input yourself, to
+                // customize the display for, e.g., NER.
+                const original = reducedInput["original"];
+                const formatted_original = reducedInput["formatted_original"];
```
formattedOriginal

joelgrus commented on demo/src/components/InputReduction.js:
```diff
@@ -0,0 +1,124 @@
+                // customize the display for, e.g., NER.
+                const original = reducedInput["original"];
+                const formatted_original = reducedInput["formatted_original"];
+                var internal_text = [];
```
let internalText = []

joelgrus commented on demo/src/components/InputReduction.js:
```diff
@@ -0,0 +1,124 @@
+                } else {
+                    reducedInput["reduced"].forEach(reduced => {
+                        const no_reduction = JSON.stringify(original) === JSON.stringify(reduced);
+                        const [original_colored, reduced_colored] = colorizeTokensForInputReductionUI(original, reduced);
```
camelCase

joelgrus commented on demo/src/components/Model.js:
```diff
@@ -65,6 +71,42 @@ class Model extends React.Component {
+        return response.json();
+      }).then((json) => {
+        const stateUpdate = { ...this.state }
+        stateUpdate['interpretData'] = Object.assign({}, { [interpreter]: json }, stateUpdate['interpretData'])
```
I probably would have done ` = {...stateUpdate['interpretData'], [interpreter]: json}` which feels cleaner to me

joelgrus commented on demo/src/components/Model.js:
```diff
@@ -65,6 +71,42 @@ class Model extends React.Component {
+
+    attackModel(inputs, attacker, name_of_input_to_attack, name_of_grad_input) {
+      const { apiUrlAttack } = this.props
+      fetch(apiUrlAttack(Object.assign(inputs, {attacker}, {name_of_input_to_attack}, {name_of_grad_input})), {
```
same comment {...inputs, attacker, etc }

joelgrus commented on demo/src/components/Model.js:
```diff
@@ -65,6 +71,42 @@ class Model extends React.Component {
+      })
+    }
+
+    attackModel(inputs, attacker, name_of_input_to_attack, name_of_grad_input) {
```
`inputToAttack`, `gradInput`

joelgrus commented on demo/src/components/Model.js:
```diff
@@ -65,6 +71,42 @@ class Model extends React.Component {
+        return response.json();
+      }).then((json) => {
+        const stateUpdate = { ...this.state }
+        stateUpdate['attackData'] = Object.assign({}, { [attacker]: json }, stateUpdate['attackData'])
```
same comment

joelgrus commented on demo/src/components/Saliency.js:
```diff
@@ -0,0 +1,170 @@
+      title = 'Simple Gradients Visualization'
+      description = <p> See saliency map interpretations generated by <a href="https://arxiv.org/abs/1312.6034" target="_blank" rel="noopener noreferrer">visualizing the gradient</a>. </p>
+    }
+    else if (interpreter === IG_INTERPRETER){
```
on previous line

joelgrus commented on demo/src/components/Saliency.js:
```diff
@@ -0,0 +1,170 @@
+      title = 'Integrated Gradients Visualization'
+      description = <p> See saliency map interpretations generated using <a href="https://arxiv.org/abs/1703.01365" target="_blank" rel="noopener noreferrer">Integrated Gradients</a>.</p>
+    }
+    else if (interpreter === SG_INTERPRETER){
```
on previous line

joelgrus commented on demo/src/components/Saliency.js:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    // Add indices so we can keep track after sorting
+    let indexedTokens = tokensWithWeights.map((obj, idx) => Object.assign({}, obj, {idx}))
```
{...obj, idx}

joelgrus commented on demo/src/components/Saliency.js:
```diff
@@ -0,0 +1,170 @@
+    const { interpretData, inputTokens, inputHeaders, interpretModel, requestData, interpreter } = this.props
+    const [title, description] = getHeaders(interpreter)
+
+    const run_button = <button
```
same comment as before

joelgrus commented on demo/src/components/Saliency.js:
```diff
@@ -0,0 +1,170 @@
+                          Interpret Prediction
+                       </button>
+
+    var display_text = '';
```
let

joelgrus commented on demo/src/components/Saliency.js:
```diff
@@ -0,0 +1,170 @@
+    if (interpretData === undefined) {
+      display_text = <div><p style={{color: "#7c7c7c"}}>Press "interpret prediction" to show the interpretation.</p>{run_button}</div>
+    } else {
+      const saliency_maps = [];
```
saliencyMaps

joelgrus commented on demo/src/components/Saliency.js:
```diff
@@ -0,0 +1,170 @@
+        // the tokens highlighted based on their top values
+        const colorMap = this.colorize(tokenWeights, topKIdx)
+        const k = i in this.state.topK ? this.state.topK[i] : this.state.topK.all
+        const saliency_map = (
```
saliencyMap

joelgrus commented on demo/src/components/demos/NamedEntityRecognition.js:
```diff
@@ -173,7 +186,70 @@ const TokenSpan = ({ token }) => {
 }
 
-const Output = ({ responseData }) => {
+const get_grad_data = (instances, num_grads) => {
```
getGradData

joelgrus replied:
numGrads

joelgrus commented on demo/src/components/demos/NamedEntityRecognition.js:
```diff
@@ -173,7 +186,70 @@ const TokenSpan = ({ token }) => {
+const SaliencyMaps = ({interpretData, tokens, relevantTokens, interpretModel, requestData}) => {
+  var simple_grad_data = undefined;
+  var integrated_grad_data = undefined;
+  var smooth_grad_data = undefined;
```
let + camelCase

joelgrus commented on demo/src/components/demos/ReadingComprehension.js:
```diff
@@ -128,10 +140,72 @@ const ArithmeticEquation = ({numbers}) => {
 }
 
-const AnswerByType = ({requestData, responseData}) => {
+const get_grad_data = ({ grad_input_1, grad_input_2 }) => {
```
camelCase

joelgrus commented on demo/src/components/demos/ReadingComprehension.js:
```diff
@@ -128,10 +140,72 @@ const ArithmeticEquation = ({numbers}) => {
+const SaliencyMaps = ({interpretData, question_tokens, passage_tokens, interpretModel, requestData}) => {
+  var simple_grad_data = undefined;
+  var integrated_grad_data = undefined;
+  var smooth_grad_data = undefined;
```
let

joelgrus commented on demo/src/components/demos/ReadingComprehension.js:
```diff
@@ -128,10 +140,72 @@ const ArithmeticEquation = ({numbers}) => {
+}
+
+const Attacks = ({attackData, attackModel, requestData}) => {
+  var hotflip_data = undefined;
```
let hotflipData

matt-gardner commented:
Thanks! On the whitespace, eslint gives no new errors.  I saw some files that had two-space indentation, and some that had four-space indentation, and wasn't really sure what to do.  I also wasn't really sure what to do with long xml tokens, so the wrapping is probably odd.  If there's some kind of style guide or auto formatter for this, pointers would be nice.

joelgrus commented:
I don't care about 2 vs 4, I meant more places where the indentation doesn't line up, e.g.

https://github.com/allenai/allennlp-demo/pull/262/files#diff-3f41e546893dc64b71aaacad12cad815R210

matt-gardner commented:
Hmm, that's annoying.  And that's in python code.  Is pylint not being run during CI?

joelgrus commented:
oh, possibly it's not on this repo 😬

matt-gardner commented:
Ok, I believe I've fixed everything you said.  PTAL.  Some snake case variables need to remain, because they are auto expanded arguments from json objects that are created in python with snake case.  But I simplified the API urls considerably (thanks for the suggestion), and cleaned up other stuff.  I locally tested the same four models, and it looks like they all work.

joelgrus commented on demo/src/components/demos/ReadingComprehension.js:
```diff
@@ -128,10 +140,72 @@ const ArithmeticEquation = ({numbers}) => {
   if(requestData && responseData) {
     const { passage, question } = requestData;
-    const { answer } = responseData;
+    const { answer, question_tokens, passage_tokens } = responseData;
```
what you can do in this case is rename as part of the destructuring:

```
const { answer, question_tokens: questionTokens, passage_tokens: passageTokens } = responseData;
```

https://wesbos.com/destructuring-renaming/
